### DISCUSSION TITLE
Chroma DB Indexes: Not exactly what you might expect...

### DISCUSSION THREAD

**Discussion by johnbrisbin** [imartinez/privateGPT#563] [Ideas]
Chroma DB is the underlying vector database used by privateGPT and it automatically creates an index of the embeddings as they are inserted during ingestion. Without that index or should it become corrupted or out of sync with the embedding, your queries will fail. They may fail obviously or mysteriously.

Contrary to the way Chroma DB is generally described, once you have specified a persistent directory on disk to store your database, Chroma DB writes to the index files continuously during ingestion, at the same time keeping the database contents in memory and only writing them to disk when the ingestion is complete (main branch) or when a checkpoint occurs (my fork).

This has a couple of consequences:
1. If a number of files are ingested after the last time the DB is persisted and then the operation is aborted before completion, their is high likelihood that the DB on disk and indexes will be out of sync. The index writer in Chroma may prevent that from being a serious problem however.
2.  If the ingest app crashes while the indexes are being updated they may be corrupted, making the database inoperable until the index problem is fixed.

Case #2 is the most worrisome since you can delete the whole DB folder (and lose a lot of ingestion time) or delete the Index folder within the DB folder which will allow the use of the DB but it will only index the files that are ingested afterwards.

There is a fix, however.
Add the following lines to the ingestion code just after the Chroma db is opened and before anything is inserted:

          if db._collection.count() and not os.path.isdir(os.path.join(persist_directory, "index")):
                print("Recreating database index. This may require a great deal of memory.")
                db._collection.create_index()

Then, if the user deletes the index folder, it will be automatically reindexed the next time the db is used, restoring full functionality with no loss of effort (though the reindexing may take a while on a slow disk or large database)
This code uses two methods that are part of the standard Chroma API, but not exposed explicitly by langchain (because of their lowest common denominator support for vector databases). It checks to see if embeddings have been inserted and if the 'index' folder is missing then creates the replacement index using the existing data already ingested.

**Comment by borisko123** [comment]
great explanation!
